### PR TITLE
Candidates

### PR DESCRIPTION
**Pull request by serg-kozhemyakin**
added new plugin that will display configured search engines in popup menu, and in new mode with search engine selection. will show menu and mode only if something exists in primary selection.

### CODE REVIEW

File: searchmenu/init.lua [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+-- View search engines in list
+new_mode("search-menu", {
+            enter = function (w)
+               local rows = {{ "Title", " URI", title = true }}
+               local selection = luakit.selection.primary
+               local search_engines = settings.get_setting_for_view(w.view, "window.search_engines")
+               if search_engines then
+                  for name, uri in pairs(search_engines) do
+                     local _title = lousy.util.escape(name)
+                     local _uri = lousy.util.escape(string.format(uri, selection))
```
Comment (aidanholm) [inline searchmenu/init.lua]: This needs to use `luakit.uri_encode()` for `selection`, otherwise it searches for the full url if the selection has spaces.

**Comment by serg-kozhemyakin** [comment]
i'm complete noob in lua so maybe you can tell me -- is there any source indenter for it that you using for formatting sources?

**Comment by aidanholm** [comment]
I use vim myself, but this seems to work well: http://blackmiaool.com/lua-beautify/ you can just use it on the website, no need to download

File: searchmenu/init.lua [hunk 2]
```diff
@@ -0,0 +1,132 @@
+-------------------------------------------------------------------------------
+--                                                                           --
+-- Search mode: add to popup menu list of all configured search egines       --
+--                                                                           --
+-------------------------------------------------------------------------------
+
+local lousy = require("lousy")
+local webview = require ("webview")
+local msg = require( "msg")
```
Comment (aidanholm) [inline searchmenu/init.lua]: `msg` isn't used, so this line should be removed

File: searchmenu/init.lua [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+local window = require ("window")
+local settings = require ("settings")
+local luakit = require ("luakit")
+local binds, modes = require("binds"), require("modes")
+local add_binds, add_cmds = modes.add_binds, modes.add_cmds
+local menu_binds = binds.menu_binds
+local new_mode = require("modes").new_mode
+
+local fav_icons = {}
```
Comment (aidanholm) [inline searchmenu/init.lua]: same here

File: searchmenu/init.lua [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            end
+            local item = string.format("Search '%s' in:", text)
+            menu[n+1] = {}
+            menu[n+1][1] = item
+            menu[n+1][2] = submenu
+        end
+    end
+end
+
+webview.add_signal("init",
```
Comment (aidanholm) [inline searchmenu/init.lua]: This block should be indented like this:
```
webview.add_signal("init", function (view)
    view:add_signal("populate-popup", populate_search_menu)
end)
```

**Comment by serg-kozhemyakin** [comment]
seems squashing pull request into one commit is not so trivial task (at least for me). so let me close this pull request and make new fresh one

**Pull request closed without merging**